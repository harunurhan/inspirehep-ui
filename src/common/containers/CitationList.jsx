import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { List } from 'immutable';

import fetchCitations from '../../actions/citations';
import ListWithPagination from '../components/ListWithPagination';
import ContentBox from '../components/ContentBox';
import CitationItem from '../components/CitationItem';

export const PAGE_SIZE = 25;

class CitationList extends Component {
  static renderCitationItem(citation) {
    return (
      <CitationItem key={citation.get('control_number')} citation={citation} />
    );
  }

  constructor(props) {
    super(props);

    this.onPageChange = this.onPageChange.bind(this);
  }

  componentDidMount() {
    const { pidType, recordId, dispatch } = this.props;
    dispatch(
      fetchCitations(pidType, recordId, { page: 1, pageSize: PAGE_SIZE })
    );
  }

  onPageChange(page) {
    const { pidType, recordId, dispatch } = this.props;
    dispatch(fetchCitations(pidType, recordId, { page, pageSize: PAGE_SIZE }));
  }

  render() {
    const { loading, total, citations } = this.props;
    return (
      <ContentBox title={`Citations (${total})`} loading={loading}>
        {total > 0 && (
          <ListWithPagination
            renderItem={CitationList.renderCitationItem}
            pageItems={citations}
            onPageChange={this.onPageChange}
            total={total}
            loading={loading}
            pageSize={PAGE_SIZE}
          />
        )}
      </ContentBox>
    );
  }
}

CitationList.propTypes = {
  total: PropTypes.number.isRequired,
  citations: PropTypes.instanceOf(List).isRequired,
  loading: PropTypes.bool.isRequired,
  dispatch: PropTypes.func.isRequired,
  pidType: PropTypes.string.isRequired,
  recordId: PropTypes.number.isRequired,
};

const stateToProps = state => ({
  loading: state.citations.get('loading'),
  citations: state.citations.get('data'),
  total: state.citations.get('total'),
});

const dispatchToProps = dispatch => ({ dispatch });

export default connect(stateToProps, dispatchToProps)(CitationList);
