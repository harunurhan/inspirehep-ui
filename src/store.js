import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import createHistory from 'history/createBrowserHistory';
import { routerMiddleware } from 'react-router-redux';
/* eslint-disable import/no-extraneous-dependencies */
import { createLogger } from 'redux-logger';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
/* eslint-disable import/no-extraneous-dependencies */

import reducers from './reducers';
import http from './common/http';
import queryParamsParserMiddleware from './middlewares/queryParamsParser';
import persistUserStateMiddleware, {
  reHydrateRootStateWithUser,
} from './middlewares/persistUserState';

export const thunkMiddleware = thunk.withExtraArgument(http);

export const history = createHistory();
const reduxRouterMiddleware = routerMiddleware(history);

const getMiddleware = () => {
  if (process.env.NODE_ENV === 'production') {
    return applyMiddleware(
      reduxRouterMiddleware,
      queryParamsParserMiddleware,
      persistUserStateMiddleware,
      thunkMiddleware
    );
  }
  return applyMiddleware(
    reduxRouterMiddleware,
    queryParamsParserMiddleware,
    persistUserStateMiddleware,
    thunkMiddleware,
    createLogger()
  );
};

export const store = createStore(
  reducers,
  reHydrateRootStateWithUser(),
  composeWithDevTools(getMiddleware())
);
