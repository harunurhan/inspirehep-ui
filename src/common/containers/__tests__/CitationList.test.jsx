import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import CitationList, { PAGE_SIZE } from '../CitationList';
import { getStoreWithState, getStore } from '../../../fixtures/store';
import fetchCitations from '../../../actions/citations';

jest.mock('../../../actions/citations');

describe('CitationList', () => {
  beforeAll(() => {
    fetchCitations.mockReturnValue(async () => {});
  });

  it('renders with state from store', () => {
    const store = getStoreWithState({
      citations: fromJS({
        loading: false,
        data: [
          {
            control_number: 170,
          },
        ],
        total: 1,
      }),
    });
    const wrapper = shallow(
      <CitationList pidType="test" recordId={123} store={store} />
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  it('calls fetchCitations on mount for first page', () => {
    const store = getStore();
    shallow(<CitationList pidType="test" recordId={123} store={store} />);
    expect(fetchCitations).toHaveBeenCalledWith('test', 123, {
      page: 1,
      pageSize: PAGE_SIZE,
    });
  });

  it('calls fetchCitations on page change', () => {
    const store = getStore();
    const wrapper = shallow(
      <CitationList pidType="test" recordId={123} store={store} />
    ).dive();
    const page = 2;
    wrapper.instance().onPageChange(page);
    expect(fetchCitations).toHaveBeenCalledWith('test', 123, {
      page,
      pageSize: PAGE_SIZE,
    });
  });

  it('does not render if total <= 0', () => {
    const store = getStoreWithState({
      citations: fromJS({
        loading: false,
        data: [
          {
            control_number: 170,
          },
        ],
        total: 0,
      }),
    });
    const wrapper = shallow(
      <CitationList store={store} pidType="test" recordId={123} />
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});
