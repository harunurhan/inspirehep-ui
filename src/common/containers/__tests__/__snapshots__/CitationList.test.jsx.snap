// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitationList does not render if total <= 0 1`] = `
<ContentBox
  actions={null}
  loading={false}
  title="Citations (0)"
/>
`;

exports[`CitationList renders with state from store 1`] = `
<ContentBox
  actions={null}
  loading={false}
  title="Citations (1)"
>
  <ListWithPagination
    loading={false}
    onPageChange={[Function]}
    pageItems={
      Immutable.List [
        Immutable.Map {
          control_number: 170,
        },
      ]
    }
    renderItem={[Function]}
    title={null}
    total={1}
  />
</ContentBox>
`;
